src/Components/Blazor/Blazor/src/Hosting/WebAssemblyHost.cs

```diff
@@ -5,94 +5,166 @@
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Blazor.Rendering;
+using Microsoft.Extensions.Configuration;
 using Microsoft.Extensions.DependencyInjection;
-using Microsoft.JSInterop;
+using Microsoft.Extensions.Logging;
 
 namespace Microsoft.AspNetCore.Blazor.Hosting
 {
-    internal class WebAssemblyHost : IWebAssemblyHost
+    /// <summary>
+    /// A host object for Blazor running under WebAssembly. Use <see cref="WebAssemblyHostBuilder"/>
+    /// to initialize a <see cref="WebAssemblyHost"/>.
+    /// </summary>
+    /// <remarks>
+    /// Calling <see cref="RunAsync"/> will result in disposal of the host object. If the host is going
+    /// to used to run an application, then it is not necessary to manually dispose the host.
+    /// </remarks>
+    public sealed class WebAssemblyHost : IDisposable, IAsyncDisposable
     {
-        private readonly IJSRuntime _runtime;
-
-        private IServiceScope _scope;
+        private readonly IServiceScope _scope;
+        private readonly IServiceProvider _services;
+        private readonly IConfiguration _configuration;
+        private readonly RootComponentMapping[] _rootComponents;
+
+        // NOTE: the host is disposable because it OWNs references to disposable things.
+        //
+        // However, the way that users primarily interact with the host is by calling RunAsync().
+        // Hosts are single-use, and once RunAsync() is called, it can't be called again for the same object.
+        // Therefore we dispose the host inside RunAsync().
```

Comment:
TBH I haven't fully understood why it's desirable to `Dispose` inside `RunAsync` instead of letting external code (unit tests, etc.) call `Dispose` manually.

I could understand why it's valuable to store a flag to make sure `RunAsync` is only called once though.

Reply:
Yeah this is one of those things that's pretty bizarre. None of the answers are going to be satisfying. 

For instance if we *don't* dispose the host inside the `RunAsync`, would we add a using?

I think I'm with you, and I'm going to just take this out because it's a confusing choice. It matters not (in the current design).